### PR TITLE
[SHAPI-861] New demographic field definitions and v3.1 schema update.

### PR DESCRIPTION
**Pull request by amanongithub7**
## Link to JIRA
[SHAPI-861](https://clever.atlassian.net/browse/SHAPI-861)

## Overview
Files I modified: `full-v3.yml` and `v3/generate.go`
Files auto-generated/auto-updated: `v3.1.yml`, `v3.1-client.yml`, `v3.1-events.yml`

Some App Partners use the API schemas provided here to generate clients manually or automatically. 
We are introducing new demographic fields in our API v3.1 Student responses.

This PR introduces these demographic fields to `full-v3.yml` which are filtered out during generation of `v3.0` ymls but not in `v3.1` yml by `v3/generate.go`.

The list of demographic fields that App Partners can receive using API v3.1 is:
- `disability`
- `gifted_status`
- `home_language_code`
- `preferred_name`
- `section_504_status`

## Testing
I ran `make generate VERSION=3` to generate the API schema files for `v3.0` and `v3.1` and have included the files in this commit. 

Edit: `frl_status` should have been included in `v3.0` schema previously but wasn't. Tracking the discrepancies and needed updates in 3.0 in this JIRA ticket: https://clever.atlassian.net/browse/SHAPI-961

## Rollout
Immediate.

[SHAPI-861]: https://clever.atlassian.net/browse/SHAPI-861?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: v3/generate.go [hunk 1]
```diff
@@ -36,23 +37,29 @@ func Generate() {
 	}
 
 	for _, versionStr := range versionStrs {
-		clientBytes, err := generateClientYml(swagger, versionStr)
+		swaggerCopy := duplicateMap(swagger)
+
+		if versionStr == "v3.0" {
+			deleteV31Definitions(swaggerCopy)
```
Comment (amanongithub7) [inline v3/generate.go:43]: I added 2 definitions to `full-v3.yml` - `Disability` and `PreferredName` - that are not required in API v3.0 schema. I've added logic here for each version to use a unique swagger copy to delete any definitions if required before YAML generations.

File: full-v3.yml [hunk 2]
```diff
@@ -1870,14 +1950,6 @@ definitions:
         - "Y"
         - "N"
         - ""
-      home_language_name:
-        type: string
-        x-nullable: true
-        x-validation: true
```
Comment (fang-chang) [inline full-v3.yml:1876]: This is confirmed that we want to remove this?

File: v3/generate.go [hunk 3]
```diff
@@ -62,6 +69,37 @@ func Generate() {
 	}
 }
 
+// duplicateMap creates a deep copy of the provided map
+func duplicateMap(m map[interface{}]interface{}) map[interface{}]interface{} {
```
Comment (fang-chang) [inline v3/generate.go:73]: Looks fine, but wondering if there's already an existing utility method out there

**Review by fang-chang — APPROVED** [review]

File: full-v3.yml (hunk 2 above)
Comment (amanongithub7) [reply full-v3.yml:1876]: Yes. We're packing the `home_language` field with the name of the language so this field is redundant. 
The enum of `home_language` is different for v3.0 and v3.1 if you look at the changes to `v3/generate.go`'s `modifyDefinitions` function.
`home_language_name` was introduced during an intern project in Summer '22.

File: v3/generate.go (hunk 3 above)
Comment (amanongithub7) [reply v3/generate.go:73]: I can't seem to find any deep copy functions in public libraries. 
The standard library has [map.Copy](https://pkg.go.dev/maps@master#Copy) and [map.Clone](https://pkg.go.dev/maps@master#Clone) but these perform assignment copies or shallow copies. Have you come across one?

Comment (fang-chang) [reply v3/generate.go:73]: Hmm, maybe https://stackoverflow.com/a/67139854 but comments say that performance is way worse. Having your own deep copy method is fine. I was just surprised that we haven't done this before

File: full-v3.yml (hunk 2 above)
Comment (amanongithub7) [reply full-v3.yml:1876]: Attaching a [link here of our demographic fields](https://docs.google.com/spreadsheets/d/15oeoRLfpWYJmxEhDl978I0j-o39WJY-5rsp49LzTqaI/edit#gid=0) for this project for your convenience.

**Pull request merged**